File: ABM-SIM-DOTNET/src/taker/TChartist.cs
namespace ABMSIM {
    public class TChartist : IAgent {
        private float forgettingFactor;
        private float EWMA;
        public TChartist(float forgettingFactor, float intitalMidPrice) : base() {
            this.forgettingFactor = forgettingFactor;
            EWMA = intitalMidPrice;
        }

        private float getEWMA(float midPrice) {
            EWMA = EWMA + (forgettingFactor) * (EWMA - midPrice);
            return EWMA;
        }

        private float getVolumeConstant(float midPrice) {
            if (Math.Abs(EWMA - midPrice) <= (midPrice) * (2f)) {
                return 20f;
            } else if (Math.Abs(EWMA - midPrice) > (midPrice) * (2f)) {
                return 50f;
            } else {
                return 0f;
            }
        }

        public MarketOrder? Action(Moment m) {
            float EWMA = getEWMA(m.midPrice);
            OrderTypes orderType;
            if (EWMA < m.midPrice - (0.5f) * (m.spread)) {
                orderType = OrderTypes.BUY;
            } else if (EWMA > m.midPrice + (0.5f) * (m.spread)) {
                orderType = OrderTypes.SELL;
            } else {
                return null;
            }
            int volume = (int)Util.calculatePowerLawVolume(orderType == OrderTypes.BUY, getVolumeConstant(m.midPrice), m);
            return new MarketOrder(orderType, volume, EWMA);
        }
    }
}

File: abm-sim-js/src/sim/agents/taker/TChartist.js
import MarketOrder, { OrderType } from "../../MarketOrder.js"
import {calculatePowerLawVolume} from '../../Util.js'

export default class TChartist {
  forgettingFactor
  EWMA
  constructor(forgettingFactor, EWMA) {
    this.forgettingFactor = forgettingFactor
    this.EWMA = EWMA
  }

  getEWMA(midPrice) {
    this.EWMA = this.EWMA + (this.forgettingFactor) * (this.EWMA - midPrice)
    return this.EWMA
  }

  getVolumeConstant(midPrice) {
    if (Math.abs(this.EWMA - midPrice) <= (midPrice) * 2) {
      return 20;
    } else if (Math.abs(this.EWMA - midPrice) > (midPrice) * 2) {
      return 50
    } else {
      return 0
    }
  }

  Action(moment) {
    let EWMA = this.getEWMA(moment.midPrice)
    let orderType
    if (EWMA < moment.midPrice - (0.5 * moment.spread)) {
      orderType = OrderType.BUY
    } else if (EWMA > moment.midPrice - (0.5 * moment.spread)) {
      orderType = OrderType.SELL
    } else {
      return null
    }
    const volume = calculatePowerLawVolume(orderType === OrderType.BUY, this.getVolumeConstant(moment.midPrice), moment)
    return new MarketOrder(orderType, Math.floor(volume), EWMA)
  }
}